docs/nebula-design-concepts.md

```diff
@@ -0,0 +1,216 @@
+pointing from team to serve, the properties are start_year(int) and end_year(int).
+
+It should be noted that there may be **multiple edges of the same or different types
+at the same time** between the start (src)and end vertex (dst).
+
+
+
+## Graph partition
+
+![image][example2]
+
+Since the number of vertices in a super large-scale relational network can reach
+to dozen billions or trillions, and the number of edges can even be dozen trillions,
+the storage of both will exceed the capacity of a general server. Therefore, a
+method is needed to cut the graph elements and store them on different logical partitions. Nebula Graph conducts **edge splitting** to make partitions, and its default strategy is Hash. The number of partitions is static and cannot be changed.
+
+
+## Data model
+
+![image][example3]
+
+In Nebula Graph, each vertex is modeled as key-values, then stored to the corresponding partition after hashing according to its vertexID.
+
+An edge is modeled as two separate key-values in Nebula Graph, i.e. out-key and
+in-key, respectively. The out-key and the edge's corresponding starting point are stored in one partition, while the in-key and end point are stored in another one.
+
+![image][example4]
+
+## Architecture
+
+Nebula Graph includes four modules: storage layer, metadata service, computation layer
+and client console.
+
+![image][example5]
+
+### Storage layer
+
+The corresponding process of storage layer is nebula-storaged, and its core is the
+distributed Key-value Storage based on RAFT protocol. Currently it supports
+multiple storage engines like Rocksdb and HBase, etc.
+
+Raft protocol ensures data consistency  by leader/follower mechanism. Based on that, Nebula Storage makes the following optimizations:
+
+- Parallel Raft
+
+      Partitions of same ID from multiple machines form a raft group. And the concurrent operations are implemented with multiple sets of Raft groups.
+
+- Write Path & batch
+
+      In Raft protocol, multi-machine synchronization depends on log ID orders, which
+  lead to low throughput. Nebula achieves high throughput by batch and out-of-order
+  submission.
+
+- Copy learner based on synchronism
+
+      When added to cluster, new machines will be tagged as learner and pull data
+  from leader/follower synchronically. The learner will be tagged as follower and
+  participate in Raft protocol when it catches up leader.
+
+- Load-balance
+
+      For machines with high stress, migrate the partition they serve to machines
+  with  low stress for better load balancing.
+
+![image][example6]
+
+### Metadata service
+
+The process corresponding to the Meta service is nebula-metad, and its main functions are:
+
+-  User management
+
+      Nebula Graph's user system includes God user, Admin, User, and Guest, each is different in operating privileges.
+
+- Cluster configuration management
+
+      Support online and offline new servers.
+
```

Comment:
Metadata service manages the servers and partitions in the cluster, e.g. records location of the partitions, receives Heartbeat from servers, etc. It balances the partitions and manages the communication traffic in case of server failure.